patches/digest-0.0.1.2.patch

```diff
@@ -0,0 +1,353 @@
++     *  The following logic has come from RFC1952.
++     */
++    private int v = 0;
++    private static int[] crc_table = null;
++    static {
++        crc_table = new int[256];
++        for (int n = 0; n < 256; n++) {
++            int c = n;
++            for (int k = 8;  --k >= 0; ) {
++                if ((c & 1) != 0)
++                    c = 0xedb88320 ^ (c >>> 1);
++                else
++                    c = c >>> 1;
++            }
++            crc_table[n] = c;
++        }
++    }
++
++    public void update(byte b) {
++        int c = ~v;
++        c = crc_table[(c^b)&0xff]^(c >>> 8);
++        v = ~c;
++    }
++
++    public void reset(long vv){
++        v = (int)(vv&0xffffffffL);
++    }
++
++    public long getValue(){
++        return (long)(v&0xffffffffL);
++    }
++}
+diff --git a/java/Digest.java b/java/Digest.java
+new file mode 100644
+index 0000000..3827b20
+--- /dev/null
++++ b/java/Digest.java
+@@ -0,0 +1,28 @@
++package eta.digest;
++
++import java.nio.ByteBuffer;
++
++/* NOTE: Hard limit on ByteBuffer API: 2^32 - 1 because of reliance on
++         signed integer type everywhere.
++*/
++
++public class Digest {
++    public static long calc_crc32(int seed, ByteBuffer buf) {
++        CRC32 checksum = new CRC32();
++        checksum.reset(seed);
++        while (buf.hasRemaining()) {
++            byte b = buf.get();
++            checksum.update(b);
++        }
++        return checksum.getValue();
++    }
++
++    public static long calc_adler32(int seed, ByteBuffer buf) {
++        Adler32 checksum = new Adler32();
++        checksum.reset(seed);
++        while (buf.hasRemaining()) {
++            checksum.update(buf.get());
++        }
++        return checksum.getValue();
++    }
++}
+diff --git a/java/Utils.java b/java/Utils.java
+new file mode 100644
+index 0000000..c8b9910
+--- /dev/null
++++ b/java/Utils.java
+@@ -0,0 +1,17 @@
++package eta.digest;
++
++import java.nio.ByteBuffer;
++
++/* NOTE: Hard limit on ByteBuffer API: 2^32 - 1 because of reliance on
++         signed integer type everywhere.
++*/
++
++public class Utils {
++    public static long calc_crc32(int seed, ByteBuffer buf) {
++        return Digest.calc_crc32(seed, buf.duplicate());
++    }
++
++    public static long calc_adler32(int seed, ByteBuffer buf) {
++        return Digest.calc_adler32(seed, buf.duplicate());
++    }
++}
+diff --git a/java/eta-0.0.9/Utils.java b/java/eta-0.0.9/Utils.java
+new file mode 100644
+index 0000000..2d08671
+--- /dev/null
++++ b/java/eta-0.0.9/Utils.java
+@@ -0,0 +1,21 @@
++package eta.digest;
++
++import java.nio.ByteBuffer;
++
++import eta.runtime.io.MemoryManager;
++
++/* NOTE: Hard limit on ByteBuffer API: 2^32 - 1 because of reliance on
++         signed integer type everywhere. */
++
++public class Utils {
++
++    public static long calc_crc32(int seed, long address, int len) {
++        ByteBuffer buf = MemoryManager.getBoundedBuffer(address);
++        return Digest.calc_crc32(seed, buf);
```

Comment:
For some reason this isn't found under 0.0.9

Reply:
Ah, sorry. Copy & paste error. I added a fixup. If it works for you with that, I'll squash it.